projects/wvr-elements/src/lib/core/wvr-animation.service.ts

```diff
@@ -0,0 +1,107 @@
+
+  constructor(private readonly builder: AnimationBuilder) { }
+
+  registerAnimationReciever(recieverName: string, component: WvrBaseComponent): void {
+    let recievers = this.recieversRegistry.get(recieverName);
+    if (!recievers) {
+      recievers = new Array<WvrBaseComponent>();
+      this.recieversRegistry.set(recieverName, recievers);
+    }
+    recievers.push(component);
+  }
+
+  registerAnimationStates(): number {
+    const id = Math.random();
+    this.animationStates.set(id, new Map<string, boolean>());
+
+    return id;
+
+  }
+
+  triggerAnimationReciever(recieverName: string): void {
+    const recievers = this.recieversRegistry.get(recieverName);
+    if (recievers) {
+      recievers.forEach(r => r.triggerAnimations('animationTrigger'));
+    }
+  }
+
+  initializeAnimationElement(stateId, animationConfig: {}, animationRootElem): void {
+    Object.keys(animationConfig)
+      .forEach(animName => {
+        const animConf = animationConfig[animName];
+        const initializationMethod = wvrAnimationInitialization[animName];
+        if (initializationMethod && animationRootElem) {
+          const from = animConf.from ?
+                      animConf.from :
+                      wvrAnimationDefaults[animName].from;
+          initializationMethod(this.animationStates.get(stateId), from, animationRootElem.nativeElement);
+        }
+      });
+  }
+
+  compileAnimation(stateId, animationName, value): AnimationMetadata | Array<AnimationMetadata> {
+
+    const a = wvrAnimations[animationName];
+
+    if (!a) {
+      console.warn(`${animationName} not a known animation.`);
+
+      return undefined;
+    }
+
+    return a(this.animationStates.get(stateId), value);
+  }
+
+  selectAnimation(stateId: number, animationName: string, timing: string,
+                  to: string, from: string, animationRoot: HTMLElement): AnimationReferenceMetadata {
+
+    const animationInput: AnimationMetadata | Array<AnimationMetadata> =
+    this.compileAnimation(stateId, animationName, animationRoot);
+
+    if (animationInput) {
+      return useAnimation(animation(animationInput), {
+        params: {
+            timing,
+            to,
+            from
+          }
+      });
+    }
+  }
+
+  playAnimation(stateId: number, animationName: string, animationConfig: {}, animationRoot: HTMLElement): void {
+
+    const timing = animationConfig[animationName] ?
+                   animationConfig[animationName].timing :
+                   wvrAnimationDefaults[animationName].timing;
+    const to = animationConfig[animationName] ?
+                   animationConfig[animationName].to :
+                   wvrAnimationDefaults[animationName].to;
+    const from = animationConfig[animationName] ?
+                   animationConfig[animationName].from :
+                   wvrAnimationDefaults[animationName].from;
+
+    const a = this.selectAnimation(stateId, animationName, timing, to, from, animationRoot);
```

Comment:
Would it be better for these methods to return promise? Could avoid the null/undefined checks and extend the function result to the method consumer. It seems it would make sense to return a promise of the animation player to expose its callback, actions, and state,

Reply:
These methods should have been private. I will change `playAnimation` to return a `AnimationPlayer` which will allow for more control over the animation. This process is not asynchronous, so for now at I am not sure a promise is the best fit. The `AnimationPlayer` does have hooks into the animation lifecycle.